### PR TITLE
Fix docker cross-platform builds

### LINKED ISSUE #11081: arm64 docker images contain x86 executables
**Bug Description**
The linux/arm64 docker images for op-node:v1.7.7 and op-node:1.7.6 contain x86 binaries:
- v1.7.7: https://console.cloud.google.com/artifacts/docker/oplabs-tools-artifacts/us/images/op-node/sha256:cb55d3d031c4596a80d94928eec18a8779bb6608e1a4e3c2f39e84d162c4a086
- v1.7.6: https://console.cloud.google.com/artifacts/docker/oplabs-tools-artifacts/us/images/op-node/sha256:a1029357658a3401970883a30f8789d47e66cb0eb841ac872146403959ca15c8

**Steps to Reproduce**

On a linux/arm64 machine:
1. Pull https://us-docker.pkg.dev/oplabs-tools-artifacts/images/op-node:v1.7.7
2. Run `docker images` to get the Image ID:
   ```
   REPOSITORY                                                TAG           IMAGE ID       CREATED        SIZE
   us-docker.pkg.dev/oplabs-tools-artifacts/images/op-node   v1.7.7        2519527b9baa   3 weeks ago    68.9MB
   ```
3. Run `docker image inspect <IMAGE_ID>` to confirm the linux/arm64 image was pulled
   ```
   ...
        "Architecture": "arm64",
        "Os": "linux",
   ...
   ```
4. Launch a container of the image with shell: `docker run --rm -it <IMAGE_ID> sh`
5. Running `op-node` will give the following, which is an indication of mismatching arch:
   ```
   /usr/local/bin/op-node: line 5: syntax error: unexpected word (expecting ")")
   ```
6. Install `file` and run `file /usr/local/bin/op-node` confirms that the binary is x86:
   ```
   /usr/local/bin/op-node: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-x86_64.so.1, Go BuildID=8I1oiant_j9Oqp1IHafV/JODLNTMItMr0Ybxl5gKM/VHCHL6QTtgs_8EdG5kkx/akTeCPB0-t1x5AWSHI21, with debug_info, not stripped
   ```

**Expected behavior**
Expected `op-node` binary to be arm64.

**Environment Information:**
 - Operating System: Alpine Linux v3.18 (for the op-node v1.7.7 docker image)
 - Package Version (or commit hash): op-node v1.7.7 and v1.7.6

---

⚠️ **Notice:** Issues that do not include the following sections will be subject to closure:
- Bug Description
- Steps to Reproduce
- Environment Information

Please ensure all required sections are filled out accurately to expedite the debugging process and improve issue resolution efficiency.

### PR DESCRIPTION
**Pull request by bitwiseguy**
The docker images being pushed to docker registry by our CI do not work on arm64 machines. This PR attempts to fix that.

Fixes https://github.com/ethereum-optimism/optimism/issues/11081

### Key change to fix cross compiling issues
Removed `TARGETOS` and `TARGETARCH` declarations from the top of the dockerfile. These args both get set automatically by the docker build kit, and the global declarations in the dockerfile caused an issue where the values were not set or passed to the underlying `make` commands which invoke `go build`.

### Additional context
Also added `CGO_ENABLED=0` to docker images that are being built for cross-platform. This will cause the **build** to fail if C code is included in the go source code. This seems better than allowing the build to pass and only catching the failure further downstream when the image is pulled and attempted to run on an arm64 machine. If we want the cross-platform builds to work with C code included, we will have to spend some more effort adding appropriate `CC` and `CXX` packages to our docker builder, then allow `CGO_ENABLED=1`. I'm open to removing this part if reviewers feel this is a bad idea

### Tests

Added `check-cross-platform` ci jobs for every image published during `scheduled-docker-publish` that is built for multiple platforms (i.e. `platforms: "linux/amd64,linux/arm64"` for the `docker-build` job)

### CODE REVIEW

**Comment by fanbsb** [comment]
Please give priority to this PR, we need the arm64 image before the hardfork on op-node 1.7.7

**Comment by fanbsb** [comment]
This PR works on M1 Mac, but not on AWS Graviton instance.

Error Message
/ # op-node --help
/usr/local/bin/op-node: line 2: syntax error: unexpected "(

File: ops/docker/op-stack-go/Dockerfile [hunk 1]
```diff
@@ -165,4 +165,4 @@ CMD ["da-server"]
 
 FROM --platform=$TARGETPLATFORM $TARGET_BASE_IMAGE as op-supervisor-target
 COPY --from=op-supervisor-builder /app/op-supervisor/bin/op-supervisor /usr/local/bin/
-ENTRYPOINT ["op-supervisor"]
+CMD ["op-supervisor"]
```
Comment (bitwiseguy) [inline ops/docker/op-stack-go/Dockerfile:161]: Does changing this from ENTRYPOINT to CMD affect the op-supervisor container functionality? This change was made so that we can run the same cross-platform ci check against the container: `op-supervisor --version`. Without the change it fails due to missing required flags

File: op-node/Makefile [hunk 2]
```diff
@@ -25,8 +25,14 @@ ifeq ($(shell uname),Darwin)
 	FUZZLDFLAGS := -ldflags=-extldflags=-Wl,-ld_classic
 endif
 
+# Check if TARGETARCH is set to arm64 and adjust compilers accordingly
+ifeq ($(TARGETARCH),arm64)
+    CC := aarch64-linux-gnu-gcc
+    CXX := aarch64-linux-gnu-g++
+endif
```
Comment (bitwiseguy) [inline op-node/Makefile]: This conditional logic exists in the makefiles instead of the higher level dockerfile. Using if/else to set an `ARG` or `ENV` within a `RUN` statement in the dockerfile changes the values on that docker layer, but the altered values are not recognized by subsequent docker layers. The downside putting the logic in the makefiles is that the same code is duplicated across all relevant makefiles

File: .circleci/config.yml [hunk 3]
```diff
@@ -2093,7 +2096,14 @@ workflows:
...
     when:
-      equal: [ build_hourly, <<pipeline.schedule.name>> ]
+      and:
+        - or:
+            # Trigger on new commits
+          - equal: [ webhook, << pipeline.trigger_source >> ]
+            # Trigger on manual triggers if explicitly requested
+          - equal: [ true, << pipeline.parameters.main_dispatch >> ]
+        - not:
+            equal: [ scheduled_pipeline, << pipeline.trigger_source >> ]
```
Comment (bitwiseguy) [inline .circleci/config.yml]: Will revert this trigger change before this PR is merged. Needed to publish and run `check-cross-platform` ci job to confirm these changes are working as expected

Comment (protolambda) [reply .circleci/config.yml]: If the trigger is permissioned correctly then I don't see a reason to remove it. Seems useful to trigger it when necessary

File: .circleci/config.yml [hunk 4]
```diff
@@ -2112,7 +2126,11 @@ workflows:
           publish: true
           context:
             - oplabs-gcr
-            - slack
```
Comment (protolambda) [inline .circleci/config.yml:2115]: Why remove `- slack`? I am unfamiliar with the slack details, but is this related to cross-building? Fine if it's cleanup

File: op-batcher/Makefile [hunk 5]
```diff
@@ -24,8 +24,14 @@ ifeq ($(shell uname),Darwin)
...
 
+# Check if TARGETARCH is set to arm64 and adjust compilers accordingly
+ifeq ($(TARGETARCH),arm64)
+    CC := aarch64-linux-gnu-gcc
+    CXX := aarch64-linux-gnu-g++
+endif
+
 op-batcher:
-	env GO111MODULE=on GOOS=$(TARGETOS) GOARCH=$(TARGETARCH) go build -v $(LDFLAGS) -o ./bin/op-batcher ./cmd
+	env GO111MODULE=on GOOS=$(TARGETOS) GOARCH=$(TARGETARCH) CC=$(CC) CXX=$(CXX) CGO_ENABLED=0 go build -v $(LDFLAGS) -o ./bin/op-batcher ./cmd
```
Comment (protolambda) [inline op-batcher/Makefile]: I think we don't have to disable CGO if CC/CXX are setup correctly.

File: op-batcher/Makefile [hunk 6]
```diff
@@ -24,8 +24,14 @@ ifeq ($(shell uname),Darwin)
 	FUZZLDFLAGS := -ldflags=-extldflags=-Wl,-ld_classic
 endif
 
+# Check if TARGETARCH is set to arm64 and adjust compilers accordingly
+ifeq ($(TARGETARCH),arm64)
+    CC := aarch64-linux-gnu-gcc
+    CXX := aarch64-linux-gnu-g++
+endif
```
Comment (protolambda) [inline op-batcher/Makefile]: It would be nice to have it check amd64 and set to amd, then if-else-branch on arm64 check, and final else branch that sets CGO_ENABLED=0.

This way I am able to build an amd64 image from my arm64 machine. And we'll have a default that makes sense if people try to target other architectures.

CGO_ENABLED=0 should entirely avoid the C toolchain, and use pure-go mode (no native platform optimizations, e.g. slower signature verification fallback). We tried to just use that as default, but for some reason it was broken (probably a dependency which didn't/doesn't support pure-go compilation).

Same applies to other make files.

File: ops/docker/op-stack-go/Dockerfile (hunk 1 above)
Comment (protolambda) [reply ops/docker/op-stack-go/Dockerfile:161]: `op-supervisor` is not running in prod, or even testnet, yet. Changing the behavior to be consistent with the other containers is good. Thanks for that.

File: ops/docker/op-stack-go/Dockerfile [hunk 7]
```diff
@@ -4,17 +4,22 @@
...
 ARG TARGET_BASE_IMAGE=alpine:3.18
 
 # We may be cross-building for another platform. Specify which platform we need as builder.
-FROM --platform=$BUILDPLATFORM golang:1.21.3-alpine3.18 as builder
+FROM --platform=$BUILDPLATFORM golang:1.21.3-bullseye as builder
 
-RUN apk add --no-cache make gcc musl-dev linux-headers git jq bash
+# Install cross-compilation tools
+RUN apt-get update && apt-get install -y \
+    crossbuild-essential-arm64 \
```
Comment (protolambda) [inline ops/docker/op-stack-go/Dockerfile]: Can we add amd64, so arm users can target amd64?

File: op-batcher/Makefile (hunk 5 above)
Comment (bitwiseguy) [reply op-batcher/Makefile]: I wasn't able to get it to work without disabling CGO. I believe it is because this PR is using debian for the builder image and alpine for the target image, which use different standard c libs (glibc vs musl). Disabling CGO forces the binary to be self-contained instead of relying on the host's c library, thus avoiding the compatibility issue

Comment (protolambda) [reply op-batcher/Makefile]: If we are disabling CGO then I think the CC and CXX are completely no-op, ignored. If we are going with the no-cgo solution then we shouldn't be installing cross-compilers or debian etc. at all.

File: .circleci/config.yml (hunk 4 above)
Comment (bitwiseguy) [reply .circleci/config.yml:2115]: This was a temp change so that I didn't produce a bunch of noisy slack notifications while testing the changes in this pr

File: op-program/Makefile [hunk 8]
```diff
@@ -15,13 +15,19 @@ LDFLAGSSTRING +=-X github.com/ethereum-optimism/optimism/op-program/version.Meta
...
+endif
+
 op-program: \
 	op-program-host \
 	op-program-client \
 	op-program-client-mips
 
 op-program-host:
-	env GO111MODULE=on GOOS=$(TARGETOS) GOARCH=$(TARGETARCH) go build -v -ldflags "$(LDFLAGSSTRING)" -o ./bin/op-program ./host/cmd/main.go
+	env GO111MODULE=on GOOS=$(TARGETOS) GOARCH=$(TARGETARCH) CC=$(CC) CXX=$(CXX) CGO_ENABLED=0 go build -v -ldflags "$(LDFLAGSSTRING)" -o ./bin/op-program ./host/cmd/main.go
```
Comment (bitwiseguy) [inline op-program/Makefile]: When we are cross building op-program are we still satisfying the requirement that the build is exactly reproducible (see comment https://github.com/ethereum-optimism/optimism/pull/10346#discussion_r1589846364 for context)?

File: op-batcher/Makefile (hunk 5 above)
Comment (bitwiseguy) [reply op-batcher/Makefile]: Good point. I will remove the CC and CXX options and use CGO_ENABLED=0 instead. This will allow us to use the smaller alpine images and enable the binaries to run on amd64 and arm64 machines.

File: ops/docker/op-stack-go/Dockerfile (hunk 7 above)
Comment (bitwiseguy) [reply ops/docker/op-stack-go/Dockerfile]: No longer relevant due to decision to use CGO_ENABLED=0

File: op-node/Makefile (hunk 2 above)
Comment (bitwiseguy) [reply op-node/Makefile]: No longer relevant due to decision to use CGO_ENABLED=0 instead of CC and CXX

File: .circleci/config.yml (hunk 3 above)
Comment (bitwiseguy) [reply .circleci/config.yml]: Changed this so that it runs on the hourly job (as before) and also can be triggered manually:
```
    when:
       or:
        - equal: [ build_hourly, <<pipeline.schedule.name>> ]
        # Trigger on manual triggers if explicitly requested
        - equal: [ true, << pipeline.parameters.docker_publish_dispatch >> ]
```

File: .circleci/config.yml (hunk 4 above)
Comment (bitwiseguy) [reply .circleci/config.yml:2115]: Added slack back in

File: op-batcher/Makefile (hunk 6 above)
Comment (bitwiseguy) [reply op-batcher/Makefile]: Decided to use CGO_ENABLED=0 to allow cross compile

**Review by protolambda — APPROVED** [review]
Regarding op-program: the host part is not subject to the same reproduction and platform target restrictions as the client part

**Review by sebastianst — APPROVED** [review]

**Review by protolambda — APPROVED** [review]

**Review by zhwrd — APPROVED** [review]
